### PR TITLE
make loadQueuePeon assign multiple segments to load

### PR DESCRIPTION
**Pull request by nishantmonu51**
modify loadQueuePeon to assign multiple segments to be loaded by the historical. 
this allows the historical to use multiple threads to download segments.

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java [hunk 1]
```diff
@@ -62,4 +62,29 @@ public String toString()
...
            '}';
   }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: would fail on subclasses

Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: Actually, I don't know what the behavior of this on instantiated classes is either.

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java [hunk 2]
```diff
@@ -62,4 +62,29 @@ public String toString()
...
+      return false;
+    }
+
+    SegmentChangeRequestDrop that = (SegmentChangeRequestDrop) o;
+
+    if (!segment.equals(that.segment)) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: If the equality really is just on the segment, can the segment be acquired from a getter for this?

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestLoad.java [hunk 3]
```diff
@@ -62,4 +62,29 @@ public String toString()
...
            '}';
   }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestLoad.java]: Same comment about class!=class

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestLoad.java [hunk 4]
```diff
@@ -62,4 +62,29 @@ public String toString()
...
+      return false;
+    }
+
+    SegmentChangeRequestLoad that = (SegmentChangeRequestLoad) o;
+
+    if (!segment.equals(that.segment)) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestLoad.java]: Same comment about getter

File: server/src/main/java/io/druid/server/coordinator/LoadQueuePeon.java [hunk 5]
```diff
@@ -190,126 +193,138 @@ public void dropSegment(
...
-        } else {
-          return;
-        }
-
-        zkWritingExecutor.execute(
-            new Runnable()
-            {
-              @Override
-              public void run()
+      if (canProcessNext()) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordinator/LoadQueuePeon.java]: Can this explicitly be a concurrent queue?

**Comment by drcrallen** [comment]
From a high level I'm curious if a load queue and a drop queue actors and fewer locks would make sense instead of this map-based-meta-queue

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java (hunk 1 above)
Comment (infynyxx) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: better to use `instanceof`, IMO

``` java
if (o == null || !(o instanceof SegmentChangeRequestDrop)) {
```

File: server/src/test/java/io/druid/server/coordinator/LoadQueuePeonTest.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        jsonMapper.readValue(
+            curator.getData().decompressed().forPath(peon.pathFor(foo)),
+            DataSegmentChangeRequest.class
+        )
+    );
+    Assert.assertEquals(20L, peon.getLoadQueueSize());
+    markRequestComplete(foo);
+    Assert.assertTrue(timing.forWaiting().awaitLatch(callbackLatch));
+    Assert.assertEquals(0L, peon.getLoadQueueSize());
+    loadQueueCache.close();
```
Comment (infynyxx) [inline server/src/test/java/io/druid/server/coordinator/LoadQueuePeonTest.java]: `loadQueueCache.close()` is already being called in `tearDown()`

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java (hunk 2 above)
Comment (xvrl) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: This is all IntelliJ auto-generated code as far as I can tell. As long as we agree on whether subclasses should be accepted for `equals`, which fields to account for, and which fields can be null, I think it's reasonable to trust the generated code. We typically don't use getters anywhere else for equals methods, so unless there's a good reason to, I would leave it as is for consistency.

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java (hunk 1 above)
Comment (xvrl) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: @infynyxx @drcrallen see my comment below on the auto-generated code.

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java (hunk 2 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: Ok, I peeked through the code and it seems DataSegmentChangeRequest seems to use more or less straight up instances of the class. 

But @nishantmonu51 : can you please add a unit test where a `DataSegmentChangeRequest` for drops and loads are read from JSON and verify that the equals method works for Jackson instanced objects?

Comment (drcrallen) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: With a unit test I'm ok with keeping it consistent with other 'auto-intellij' aspects of the code.

**Comment by cheddar** [comment]
Single threaded loading of segments was done intentionally in order to simplify the ability to reason about how resources are going to be needed.  Doing this is going to cause the ability for stampedes to happen that slow down query performance.

:-1:

**Comment by drcrallen** [comment]
@cheddar : are you worried about disk-io or simple user-time contention?

**Comment by xvrl** [comment]
@nishantmonu51, @cheddar and I discussed this a little bit offline and came to the following conclusion that we think should address everyone's concerns.

The main issue we are trying to address is the ability to speedup downloading of segments when spinning up new nodes. Currently this can take tens of hours on very large clusters, because network bandwidth cannot be fully utilized, which makes adjustment of replication strategies (e.g. when adding tiers), or replacing all the hardware for an entire cluster (e.g. when AWS decides to reboot your entire cluster for patches) unnecessarily time-consuming.

We don't want parallelized segment loading could to adversely impact query performance at any cost, so we should only allow parallelization of segment loading during the initialization phase.  By default we should leave it single threaded, similar to what we do for segment cache initialization.
At query time, we should limit the number of loading threads to 1. In case, someone really wants to sacrifice 1 or 2 processing threads, and use them for loading on a 32 core machine, they should be able to do so, but we won't make that explicitly configurable, requiring overriding the config via extensions instead, to make it clear this is not something we condone.

To summarize, we should:
- make parallelization of segment loading during configurable for init phase
- default to single threaded loading during init phase to preserve the existing behavior
- have a hard-coded setting for number of loading threads post-initialization, set to 1
- allow overriding of the hard-coded setting via guice injection overrides.

Does that sound reasonable to everyone?

**Comment by cheddar** [comment]
One extra thing, on the "init phase loading", it likely might make sense to add something that allows you to fire up a historical node as a "replica" of some other historical nodes.  I.e. on first startup, it looks at what some current historical node(s) are serving and just loads itself up with those segments as fast as possible.

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java (hunk 2 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: can someone explain why suddenly equals(), hashCode() implementation is required for SegmentChangeRequest[Drop/Load]?

**Comment by nishantmonu51** [comment]
having parallel segment loading during init-phase sounds good to me as it wont affect the query performance and also decrease the node init time significantly. 
I can still see one problem with performance, when we add a new node to a tier, during its init phase queries for complete tier will slow down (as the query time is bounded by the slowest historical node).
A workaround can be to not send queries to a node until its initialized (may be decrease the priority of a node during init-phase.) 

Also, right now we do not have a well defined Lifecycle of a node and no init-phase as well,
I think the two problems of blacklisting nodes and fast init-phase may be related to node lifecycle and can be handled easily if we introduce the concept of lifecycle and state of a node.

**Comment by nishantmonu51** [comment]
@cheddar regarding having an option to start up a replica of some other historical nodes, it will solve the case of quickly replacing a node with a new node, however there are cases when we need to replicate/move data based on coordinator rules (for selective datasources and periods) across tiers and not just replicate a set of existing historical nodes. let me describe a concrete use cases - I am experimenting with a tier of hs1.8xlarge nodes, that can be used as a replica of existing tiers, and can be used to load up super old data for which performance is not critical, now the issue with these nodes is it takes order of a week to load up a node with 20T of data when the loading is single threaded. 

how about having an option to dynamically enable/disable/configure parallel loading of segments for a particular tier or set of nodes with the default to do single threaded loading to ensure that no existing behavior will change ? 
System admin can then tune the concurrency factor through coordinator console whenever required.

**Comment by cheddar** [comment]
@nishantmonu51 What if you could start up the node as a replica of *multiple* other nodes?  It could download the whole set of segments and then when the node actually shows up, the coordinator can make sure that things make sense?

That is, you bring up your hs1.8xl's set to replicate like 9 2TB nodes (for example).  That initialization happens as fast as possible during the "init" phase.  Then they join and the coordinator does its stuff?

**Comment by fjy** [comment]
@cheddar @nishantmonu51 Was there any further progress on this PR?

File: server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java (hunk 2 above)
Comment (xvrl) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: @himanshug it would be required for serde testing.

Comment (himanshug) [inline server/src/main/java/io/druid/server/coordination/SegmentChangeRequestDrop.java]: @xvrl thanks, got it.

**Comment by fjy** [comment]
@cheddar @nishantmonu51 ?

**Comment by drcrallen** [comment]
@nishantmonu51 If this isn't going to be making progress can it be closed until whatever core issue is resolved?

**Comment by nishantmonu51** [comment]
closing since having multithreaded downloading on segments causes performance to degrade.
